contracts/actions/uniswap/v3/UniSupplyV3.sol

```diff
@@ -0,0 +1,153 @@
+        uint256 amount1Desired;
+        uint256 amount0Min;
+        uint256 amount1Min;
+        uint256 deadline;
+        address from;
+    }
+
+    /// @inheritdoc ActionBase
+    function executeAction(
+        bytes[] memory _callData,
+        bytes[] memory _subData,
+        uint8[] memory _paramMapping,
+        bytes32[] memory _returnValues
+    ) public payable virtual override returns (bytes32) {
+        Params memory uniData = parseInputs(_callData);
+        
+        uniData.tokenId = _parseParamUint(uniData.tokenId, _paramMapping[0], _subData, _returnValues);
+        uniData.amount0Desired = _parseParamUint(uniData.amount0Desired, _paramMapping[1], _subData, _returnValues);
+        uniData.amount1Desired = _parseParamUint(uniData.amount1Desired, _paramMapping[2], _subData, _returnValues);
+
+        uint128 liquidity = _uniSupplyPosition(uniData);
+        return bytes32(uint256(liquidity));
+    }
+
+    /// @inheritdoc ActionBase
+    function executeActionDirect(bytes[] memory _callData) public payable override {
+        Params memory uniData = parseInputs(_callData);
+        _uniSupplyPosition(uniData);
+        
+    }
+
+    /// @inheritdoc ActionBase
+    function actionType() public pure virtual override returns (uint8) {
+        return uint8(ActionType.STANDARD_ACTION);
+    }
+
+    //////////////////////////// ACTION LOGIC ////////////////////////////
+
+    function _uniSupplyPosition(Params memory _uniData)
+        internal
+        returns(uint128 liquidity)
+    {  
+        (address token0, address token1) = _getTokenAdresses(_uniData.tokenId);
+        // fetch tokens from address
+        uint amount0Pulled = token0.pullTokensIfNeeded(_uniData.from, _uniData.amount0Desired);
+        uint amount1Pulled = token1.pullTokensIfNeeded(_uniData.from, _uniData.amount1Desired);
+
+        // approve positionManager so it can pull tokens
+        token0.approveToken(address(positionManager), amount0Pulled);
+        token1.approveToken(address(positionManager), amount1Pulled);
+
+        _uniData.amount0Desired = amount0Pulled;
+        _uniData.amount1Desired = amount1Pulled;
+
+        uint256 amount0;
+        uint256 amount1;
+        (liquidity, amount0, amount1) = _uniSupply(_uniData);
+
+        //send leftovers
+        token0.withdrawTokens(_uniData.from, sub(_uniData.amount0Desired, amount0));
+        token1.withdrawTokens(_uniData.from, sub(_uniData.amount1Desired, amount1));
+
+        logger.Log(
+                address(this),
+                msg.sender,
+                "UniSupplyV3",
+                abi.encode(_uniData, liquidity, amount0, amount1)
+            );
+
+    }
+
+    /// @dev calls positions from NonFungiblePositionManager for tokenId, and returns addresses for both tokens
+    /// @dev workaround for stack too deep error that happens because positions() returns 12 variables
+    function _getTokenAdresses(uint tokenId) internal view returns(address, address){
```

Comment:
This opens up extra complexity due to it's use of inline assembly which we generally try to avoid (unless there is no other option). The second thing to note is that this using around 10k gas, which is quite a lot for type of action.

Given that the information will be already displayed and fetched in the frontend, a much cleaner solution and gas cost effective would be to just pass pool addresses as inputs (would cost around 1k gas).